`reflection_pad1`: port to structured kernel

zyan0 opened the pull request:
Following https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

Tracking issue #55070 

Differential Revision: D27628059

ezyang commented on aten/src/ATen/native/cuda/ReflectionPad.cu:
```diff
@@ -397,17 +397,9 @@ void reflection_pad2d_backward_out_template(
-    Tensor& output) {
+TORCH_IMPL_FUNC(reflection_pad1d_out_cuda)
+(const Tensor& input, IntArrayRef padding, const Tensor& output) {
   reflection_pad1d_out_template(output, input, padding);
```
This will "work" but it will be better to inline reflection_pad1d_out_template  into the IMPL_FUNC and then delete the unnecessary shape checking code.

ezyang commented on aten/src/ATen/native/ReflectionPad.cpp:
```diff
@@ -538,6 +596,11 @@ Tensor reflection_pad1d_cpu(const Tensor& input, IntArrayRef padding) {
+TORCH_IMPL_FUNC(reflection_pad1d_out_cpu)
+(const Tensor& input, IntArrayRef padding, const Tensor& output) {
+  reflection_pad1d_out_template(output, input, padding);
+}
```
Ditto here

zyan0 replied:
reflection_pad1d_out_template is still used in reflection_pad1d_cpu for reflection_pad1d to dispatch QuantizedCPU.

ezyang reviewed (approved):
This is a good start but it would be better to now remove the redundant checks.

Commit:
Port reflection_pad1d to structured kernel (#55531)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/55531

Following https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md

Test Plan: unittests

Reviewed By: ezyang

Differential Revision: D27628059

fbshipit-source-id: 3c314e6263313d13daa08da81f76eed29b390de6
